gobblin-data-management/src/main/java/gobblin/data/management/copy/CopySource.java

```diff
@@ -246,13 +280,20 @@ public Void call() {
         while (fileSets.hasNext() && !shouldStopGeneratingWorkUnits(this.workUnitList)) {
           FileSet<CopyEntity> fileSet = fileSets.next();
           String extractId = fileSet.getName().replace(':', '_');
+
           Extract extract = new Extract(Extract.TableType.SNAPSHOT_ONLY, CopyConfiguration.COPY_PREFIX, extractId);
           List<WorkUnit> workUnitsForPartition = Lists.newArrayList();
           for (CopyEntity copyEntity : fileSet.getFiles()) {
 
             CopyableDatasetMetadata metadata = new CopyableDatasetMetadata(this.copyableDataset);
             CopyEntity.DatasetAndPartition datasetAndPartition = copyEntity.getDatasetAndPartition(metadata);
-
+            Optional<WatermarkInterval> watermarkIntervalOptional =
+                getCopyEntityWatermark(copyEntity, watermarkGenerator);
+            if (shouldSkipDueToWatermark(filterWuByWatermark,
```

Comment:
Is there a way that the `Dataset` itself can use the watermark? This is doing a late filtering, but the dataset might be able to optimize this filtering, or use the watermark to do a faster listing.